dotnet/src/AutoGen.SemanticKernel/Middleware/KernelPluginMiddleware.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// KernelPluginMiddleware.cs
+
+using System;
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using System.Threading;
+using System.Threading.Tasks;
+using AutoGen.SemanticKernel.Extension;
+using Microsoft.SemanticKernel;
+
+namespace AutoGen.SemanticKernel;
+
+/// <summary>
+/// A middleware that consumes <see cref="KernelPlugin"/>
+/// </summary>
+public class KernelPluginMiddleware : IMiddleware
+{
+    private readonly KernelPlugin _kernelPlugin;
+    private readonly FunctionCallMiddleware _functionCallMiddleware;
+    public string? Name => nameof(KernelPluginMiddleware);
+
+    public KernelPluginMiddleware(Kernel kernel, KernelPlugin kernelPlugin)
+    {
+        _kernelPlugin = kernelPlugin;
+        var functionContracts = kernelPlugin.Select(k => k.Metadata.ToFunctionContract());
+        var functionMap = kernelPlugin.ToDictionary(kv => kv.Metadata.Name, kv => InvokeFunctionPartial(kernel, kv));
+        _functionCallMiddleware = new FunctionCallMiddleware(functionContracts, functionMap, Name);
+    }
+
+    public Task<IMessage> InvokeAsync(MiddlewareContext context, IAgent agent, CancellationToken cancellationToken = default)
+    {
+        return _functionCallMiddleware.InvokeAsync(context, agent, cancellationToken);
+    }
+
+    private async Task<string> InvokeFunctionAsync(Kernel kernel, KernelFunction function, string arguments)
+    {
+        var kernelArguments = new KernelArguments();
+        var parameters = function.Metadata.Parameters;
+        var jsonObject = JsonSerializer.Deserialize<JsonObject>(arguments) ?? new JsonObject();
+        foreach (var parameter in parameters)
+        {
+            var parameterName = parameter.Name;
+            if (jsonObject.ContainsKey(parameterName))
+            {
+                var parameterType = parameter.ParameterType ?? throw new ArgumentException($"Missing parameter type for {parameterName}");
+                var parameterValue = jsonObject[parameterName];
+                var parameterObject = parameterValue.Deserialize(parameterType);
+                kernelArguments.Add(parameterName, parameterObject);
+            }
+            else
+            {
+                if (parameter.DefaultValue != null)
+                {
+                    kernelArguments.Add(parameterName, parameter.DefaultValue);
+                }
+                else if (parameter.IsRequired)
+                {
+                    throw new ArgumentException($"Missing required parameter: {parameterName}");
+                }
+            }
+        }
+        var result = await function.InvokeAsync(kernel, kernelArguments);
```

Comment:
Question, I see that this is `async` and  `function.InvokeAsync()`[ takes in a c](https://learn.microsoft.com/en-us/dotnet/api/microsoft.semantickernel.kernelfunction.invokeasync?view=semantic-kernel-dotnet#microsoft-semantickernel-kernelfunction-invokeasync-1(microsoft-semantickernel-kernel-microsoft-semantickernel-kernelarguments-system-threading-cancellationtoken))ancellationToken

Should a `cancellationToken` be passed to this ?

Reply:
Thanks for catching, obviously

Reply:
Oh,, it's actually not an easy change because function executor doesn't accept a cancellationtoken when executing.. I'll create an issue to track this